File: python/path_traversal.py
import os
from pathlib import Path

base_dir = '/var/app/files'
filename = input('Enter filename: ')
# CWE-22: no path normalization allows directory traversal (e.g. ../../etc/passwd).
base_dir_path = Path(base_dir)
file_path = base_dir_path / filename
resolved_path = file_path.resolve()
if not str(resolved_path).startswith(str(base_dir_path.resolve()) + os.sep):
    raise ValueError('Invalid path: directory traversal detected')
with open(resolved_path) as f:
    print(f.read())

File: javascript/path_traversal.js
const fs = require('fs');
const path = require('path');
const readline = require('readline');

const baseDir = '/var/app/files';
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
rl.question('Enter filename: ', (filename) => {
  // CWE-22: no path normalization — allows ../../etc/passwd traversal.
  const fullPath = path.join(baseDir, filename);
  console.log(fs.readFileSync(fullPath, 'utf8'));
  rl.close();
});
